esmvalcore/experimental/config/_config_object.py

```diff
@@ -0,0 +1,191 @@
+        self.init_session_dir(name)
+
+    def init_session_dir(self, name: str = 'session'):
+        """Initialize session.
+
+        The `name` is used to name the working directory, e.g.
+        `recipe_example_20200916/`. If no name is given, such as in an
+        interactive session, defaults to `session`.
+        """
+        now = datetime.utcnow().strftime("%Y%m%d_%H%M%S")
+        session_name = f"{name}_{now}"
+        self._session_dir = self['output_dir'] / session_name
+
+    @property
+    def session_dir(self):
+        """Return session directory."""
+        return self._session_dir
+
+    @property
+    def preproc_dir(self):
+        """Return preproc directory."""
+        return self.session_dir / 'preproc'
+
+    @property
+    def work_dir(self):
+        """Return work directory."""
+        return self.session_dir / 'work'
+
+    @property
+    def plot_dir(self):
+        """Return plot directory."""
+        return self.session_dir / 'plots'
+
+    @property
+    def run_dir(self):
+        """Return run directory."""
+        return self.session_dir / 'run'
+
+    @property
+    def config_dir(self):
+        """Return user config directory."""
+        return USER_CONFIG_DIR
+
+
+def read_config_file(config_file):
+    """Read config user file and store settings in a dictionary."""
+    config_file = Path(config_file)
+    if not config_file.exists():
+        raise IOError(f'Config file `{config_file}` does not exist.')
+
+    with open(config_file, 'r') as file:
+        cfg = yaml.safe_load(file)
+
+    return cfg
+
+
+def _load_default_config(filename: str):
+    """Load the default configuration."""
+    mapping = read_config_file(filename)
+
+    global CFG_DEFAULT
+
+    CFG_DEFAULT.update(mapping)
+
+
+def _load_user_config(filename: str, raise_exception: bool = True):
+    """Load user configuration from the given file.
+
+    The config is cleared and updated in-place.
+
+    Parameters
+    ----------
+    filename: pathlike
+        Name of the config file, must be yaml format
+    raise_exception : bool
+        Raise an exception if `filename` can not be found (default).
+        Otherwise, silently pass and use the default configuration. This
+        setting is necessary for the case where `.esmvaltool/config-user.yml`
+        has not been defined (i.e. first start).
+    """
+    try:
+        mapping = read_config_file(filename)
+        mapping['config_file'] = filename
+    except IOError:
+        if raise_exception:
+            raise
+        mapping = {}
+
+    global CFG
+
+    CFG.clear()
+    CFG.update(CFG_DEFAULT)
+    CFG.update(mapping)
+
+
+def get_user_config_location():
+    """Get the user config location by looking in the expected places.
+
+    Check if environment variable `ESMVALTOOL_CONFIG` exists, otherwise
+    use the default location in the user config dir (`~/.esmvaltool`).
+    """
+    try:
+        config_location = Path(os.environ['ESMVALTOOL_CONFIG'])
```

Comment:
I'm not too keen on supporting environmental variables, because they lead to very hard to debug situations, e.g. if people set them and then run into trouble a few years later because there was something set in their bashrc file that they completely forgot about

Reply:
I have found them super useful for all sorts of scenarios. I can remove them for now.

Reply:
Thanks